File: experiments_realworld/graphs/problem01/solutions/solution.cpp
// C++ Reference Solution - CSES 1672: Shortest Routes II
// Algorithmically equivalent to Python solution (see formal_proof.md)
// Floyd-Warshall All-Pairs Shortest Path - O(n³)

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m, q;
    if (!(cin >> n >> m >> q)) return 0;

    const long long INF = (1LL<<62);
    vector<vector<long long>> dist(n, vector<long long>(n, INF));
    for (int i = 0; i < n; ++i) dist[i][i] = 0;

    for (int i = 0; i < m; ++i) {
        int a, b; long long c;
        cin >> a >> b >> c;
        --a; --b;
        if (c < dist[a][b]) {
            dist[a][b] = c;
            dist[b][a] = c; // undirected graph
        }
    }

    // Floyd–Warshall with INF check to avoid overflow
    for (int k = 0; k < n; ++k) {
        for (int i = 0; i < n; ++i) {
            if (dist[i][k] == INF) continue;
            for (int j = 0; j < n; ++j) {
                if (dist[k][j] == INF) continue;
                long long via = dist[i][k] + dist[k][j];
                if (via < dist[i][j]) dist[i][j] = via;
            }
        }
    }

    while (q--) {
        int a, b; cin >> a >> b; --a; --b;
        long long ans = dist[a][b];
        cout << (ans >= INF ? -1 : ans) << '\n';
    }
    return 0;
}

File: experiments_realworld/graphs/problem01/solutions/solution.py
# Python Reference Solution - CSES 1672: Shortest Routes II
# Algorithmically equivalent to C++ solution (see formal_proof.md)
# Floyd-Warshall All-Pairs Shortest Path - O(n³)
# Recommended to run on PyPy for performance margin

import sys

data = list(map(int, sys.stdin.buffer.read().split()))
it = iter(data)
n = next(it); m = next(it); q = next(it)

INF = 10**18
dist = [[INF]*n for _ in range(n)]
for i in range(n):
    dist[i][i] = 0

# edges: keep the smallest among multiple
for _ in range(m):
    a = next(it) - 1
    b = next(it) - 1
    c = next(it)
    if c < dist[a][b]:
        dist[a][b] = c
        dist[b][a] = c

# Floyd–Warshall with INF check to avoid overflow
for k in range(n):
    dk = dist[k]
    for i in range(n):
        dik = dist[i][k]
        if dik == INF:
            continue
        di = dist[i]
        for j in range(n):
            dkj = dk[j]
            if dkj == INF:
                continue
            via = dik + dkj
            if via < di[j]:
                di[j] = via

# answer queries
out = []
for _ in range(q):
    a = next(it) - 1
    b = next(it) - 1
    ans = dist[a][b]
    out.append(str(-1 if ans >= INF else ans))
sys.stdout.write("\n".join(out))
